draft-normington-jsonpath-latest.xml

```diff
@@ -381,57 +381,66 @@ array-slice = [ integer ] ws %x3A ws [ integer ]
+                     <t>The resulting array is then constructed by iterating from <tt>start</tt> (inclusive)
+                        to <tt>end</tt> (exclusive), collecting array elements at index <tt>i</tt>,
+                        and incrementing the iterator <tt>i</tt> by <tt>step</tt> after each iteration. </t>
+                     <t>If <tt>len</tt> is <tt>0</tt> the resulting array is empty.</t>
+                     <t>If <tt>step &gt; 0</tt> and <tt>start &gt; end</tt> the resulting array is empty.</t>
+                     <t>If <tt>step &lt; 0</tt> and <tt>start &lt; end</tt> the resulting array is empty.</t>
+                  </section>
+                  <section numbered="false" toc="exclude">
+                     <name>Pseudocode</name>
+                     <t>If <tt>step &gt; 0</tt>, the indices of the slice are the same
+                        as the values of <tt>i</tt> iterated over by this for loop in C:</t>
                      <sourcecode type="c">
 for (i = start; i &lt; end; i = i + step) {
   ...
 }
                      </sourcecode>
-                     <t>When <tt>step</tt> is positive, the default value of <tt>start</tt> is <tt>0</tt> and the default value of <tt>end</tt> is the length of the array.
-                     </t>
-                     <t>For negative integer <tt>step</tt> and integers <tt>start</tt> and <tt>end</tt>, the array slice
-                        <tt>start:end:step</tt> consists of the indices starting at <tt>start</tt> and decreasing repeatedly
-                        by <tt>step</tt> while the resultant index is greater than <tt>end</tt>.
-                        The indices of the slice are the same as the values of i iterated over by this <tt>for</tt> loop in C:
-                     </t>
+                     <t>If <tt>step &lt; 0</tt>, the iteration will proceed backwards and the iteration test must be reversed:</t>
                      <sourcecode type="c">
 for (i = start; i &gt; end; i = i + step) {
   ...
 }
                      </sourcecode>
-                     <t>When <tt>step</tt> is negative, the default value of <tt>start</tt> is <tt>len - 1</tt> (where <tt>len</tt> is the length of the array)
-                        and the default value of <tt>end</tt> is <tt>-1</tt> (literally, not in the earlier sense of a shorthand for <tt>len - 1</tt>).
-                     </t>
-                     <t>For zero <tt>step</tt>, the array slice is empty and no elements are selected.
-                     </t>
-                  </section>       
-               </section>             
-            </section>       
-         </section>       
+                     <t>If the <tt>end</tt> bound is omitted, the iteration must terminate at 0 inclusive.
+                     This can be achieved by setting <tt>end = -1</tt>, if the underyling numeric type allows it.
+                     Alternatively the termination condition can be changed to <tt>i &gt;= 0</tt>.</t>
```

Comment:
I think we might be going into too much implementation detail here, which sets a nasty precedent for the rest of the spec. If we keep the pseudocode, I wonder if we can somehow indicate that it isn't normative.

Reply:
PTAL